File: interview/isInterleave.go
package interview

func isInterleave(s1 string, s2 string, s3 string) bool {
	if len(s1)+len(s2) != len(s3) {
		return false
	}
	f := func(i1, i2, i3 int) bool { return true }
	f = func(i1, i2, i3 int) bool {
		for i3 < len(s3) {
			if i1 == len(s1) {
				for i2 < len(s2) {
					if s2[i2] != s3[i3] {
						return false
					}
					i3++
					i2++
				}
				return true
			}
			if i2 == len(s2) {
				for i1 < len(s1) {
					if s1[i1] != s3[i3] {
						return false
					}
					i3++
					i1++
				}
				return true
			}
			if s1[i1] == s2[i2] {
				if s1[i1] != s3[i3] {
					return false
				}
				return f(i1+1, i2, i3) || f(i1, i2+1, i3)
			}
			if s1[i1] == s3[i3] {
				i3++
				i1++
			} else if s2[i2] == s3[i3] {
				i3++
				i2++
			} else {
				return false
			}
		}
		return true
	}
	return f(0, 0, 0)
}

File: algo2024/src/main/java/IsInterleave.java
public class IsInterleave {

    public boolean isInterleave(String s1, String s2, String s3) {
        if(s1.length() + s2.length() != s3.length()){
            return false;
        }
        int[] dp = new int[s1.length()+1];
        dp[0] = 1;

        for(int i = 0; i < s3.length(); i++){
            boolean flag = false;
            for(int j = dp.length-1; j >= 0; j--){
                if(dp[j] != 0){
                    dp[j] = 0;
                    if(j < s1.length() && s3.charAt(i) == s1.charAt(j)){
                        flag = true;
                        dp[j+1] = 1;
                    }
                    if(i-j < s2.length() && s3.charAt(i) == s2.charAt(i-j)){
                        flag = true;
                        dp[j] = 1;
                    }
                }
            }
            if(!flag){
                return false;
            }
        }
        return true;
    }

}
